keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+
+Note that while the initial proposal of InternalTrafficPolicy proposed a
+Local policy, it was dropped later on. This meant that now
+TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
+allow users to express a much desired use case from topologyKeys which is
+"prefer node-local, failover to same zone, then route anywhere" While this
+specific behavior is a non-goal for the initial implementation of the
+`routingPreference` field, the design allows for the potential introduction of
+such a preference in future refinements.
+
+## Goals
+
+* **Enhanced User Control:** Provide users with a knob to specify a preference that
+  MAY influence the traffic routing of a Kubernetes Service.
+
+* **Guidance for Implementations:** Offer standard routing preference values
+  that implementations can choose to recognize and support, promoting a degree
+  of commonality.
+
+* **Flexibility:** Allow implementations to interpret and support standard
+  heuristics in a manner that aligns with their capabilities and design.
+
+* **Extensibility:** Enable Kubernetes implementations to introduce and support
+  innovative routing strategies (for example, those based on topology, latency,
+  or other custom heuristics) that can evolve alongside users needs and
+  infrastructure capabilities.
+
+## Non-Goals
+
+* **Strict Routing Guarantees:** The `routingPreference` field will not
+  enforce deterministic routing paths. It serves as a mechanism for expressing
+  hints and preferences that implementations can consider when making routing
+  decisions.
+
+* **Mandatory and Uniform Implementation Support:** Kubernetes implementations
+  are not required to support all standard heuristics (e.g., `Zone`, `Spread`).
+  Even when standard heuristics are supported, their precise behavior and
+  interpretation might vary across implementations.
+
+* **Replacement of Traffic Policies:** The new field is complementary to
+  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
+  their role in enforcing strict traffic locality.
+
+* **Immediate Support for All Possible Heuristics:** The initial implementation
+  focuses on a core set of heuristics. Addition of new heuristics (like
+  `Local` for Node local preference) could be explored in future
+  refinements.
+
+## Proposal
+
+Add a new field, `routingPreference`, to the Service specification. This field
+will serve as preference or hint for the underlying implementation to consider
+while making routing decisions. It does not offer strict routing guarantees.
+
+The field will support the following initial values:
+
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `Spread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
```

Comment:
That sounds like the behavior that is currently expected from kube-proxy (EDIT: yes, basically confirmed below), and yet, by implication, it is _not_ what `Default` means? It seems to me that in past discussions Tim was pretty insistent that kube-proxy _must_ "encourage an equal distribution of traffic across endpoints" by default.

So maybe this should be the `Default`/`nil` behavior, and you should have `Auto` (?) for the "no specific routing preference" behavior.

Reply:
I think the idea here is to not completely prevent us from having smarter default behavior in the future, but I agree that the default in most cases should be `Spread`. I'd just like to hold out hope that somehow there can be a sufficiently safe "Close" mode that could be enabled by default in the future.